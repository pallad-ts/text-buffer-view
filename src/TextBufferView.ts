import {Encoding} from "./Encoding";
import {isHexString} from "./utils/isHexString";
import {isBase64} from "./utils/isBase64";
import {isBase64Url} from "./utils/isBase64Url";

export class TextBufferView {
	constructor(private buffer: ArrayBuffer) {
		if (!ArrayBuffer.isView(buffer) || buffer instanceof DataView) {
			throw new TypeError('Invalid ArrayBuffer');
		}

		Object.freeze(this);
	}

	toString(encoding: Encoding = 'utf8') {
		if (encoding === 'utf8' || encoding === 'ascii') {
			return new TextDecoder(encoding).decode(this.buffer);
		}
		return Buffer.from(this.buffer).toString(encoding);
	}

	/**
	 * Returns reference to original buffer
	 */
	get originalArrayBuffer() {
		return this.buffer;
	}

	/**
	 * Returns copy of buffer
	 */
	get arrayBuffer() {
		return this.buffer.slice(0)
	}

	static fromBuffer(buffer: Buffer) {
		return new TextBufferView(buffer);
	}

	static fromArrayBuffer(arrayBuffer: ArrayBuffer) {
		return new TextBufferView(arrayBuffer);
	}

	static fromString(input: string, encoding: Encoding) {
		if (encoding === 'utf8') {
			return new TextBufferView(new TextEncoder().encode(input));
		}

		if (encoding === 'hex') {
			if (isHexString(input)) {
				return new TextBufferView(Buffer.from(input, 'hex'));
			}
			throw new TypeError('Invalid hex string');
		}

		if (encoding === 'base64') {
			if (isBase64(input)) {
				return new TextBufferView(Buffer.from(input, 'base64'));
			}
			throw new TypeError('Invalid base64 string');
		}

		if (encoding === 'base64url') {
			if (isBase64Url(input)) {
				return new TextBufferView(Buffer.from(input, 'base64url'));
			}
			throw new TypeError('Invalid base64url string');
		}

		return new TextBufferView(Buffer.from(input, encoding));
	}
}
