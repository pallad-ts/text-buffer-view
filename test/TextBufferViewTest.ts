import {TextBufferView} from "@src/TextBufferView";

describe('TextBufferView', () => {

	describe('hex', () => {
		it('should create from string', () => {
			const input = '60cdac8b3938f188ed601cca46150196da1ff1c4';
			const buffer = TextBufferView.fromString(input, 'hex');
			expect(buffer.toString('hex')).toBe(input);
			expect(buffer).toMatchSnapshot();
		});

		it('should fail for invalid string', () => {
			expect(() => {
				TextBufferView.fromString('zzz', 'hex');
			}).toThrowError('Invalid hex string');
		});

		it('allows using empty string', () => {
			expect(TextBufferView.fromString('', 'hex').originalArrayBuffer.byteLength).toBe(0)
		})
	});

	describe('base64', () => {
		it('should create from string', () => {
			const input = 'HZB0oAfVQ890tA==';
			const buffer = TextBufferView.fromString(input, 'base64');
			expect(buffer.toString('base64')).toBe(input);
			expect(buffer).toMatchSnapshot();
		});

		it('should fail for invalid string', () => {
			expect(() => {
				TextBufferView.fromString(')(123asd', 'base64');
			}).toThrowError('Invalid base64 string');
		});

		it('allows using empty string', () => {
			expect(TextBufferView.fromString('', 'base64').originalArrayBuffer.byteLength).toBe(0)
		})
	});

	describe('base64url', () => {
		it('should create from string', () => {
			const input = 'n2f8bBHZEF-f6w';
			const buffer = TextBufferView.fromString(input, 'base64url');
			expect(buffer.toString('base64url')).toBe(input);
			expect(buffer).toMatchSnapshot();
		});

		it('should fail for invalid string', () => {
			expect(() => {
				TextBufferView.fromString(')(123asd', 'base64url');
			}).toThrowError('Invalid base64url string');
		});

		it('allows using empty string', () => {
			expect(TextBufferView.fromString('', 'base64url').originalArrayBuffer.byteLength).toBe(0)
		})
	});

	describe('ascii', () => {
		it('should create from string', () => {
			const input = 'ascii chars';
			const buffer = TextBufferView.fromString(input, 'ascii');
			expect(buffer.toString('ascii')).toBe(input);
			expect(buffer).toMatchSnapshot();
		});

		it('allows using empty string', () => {
			expect(TextBufferView.fromString('', 'ascii').originalArrayBuffer.byteLength).toBe(0)
		})
	});

	describe('utf8', () => {
		it('should create from string', () => {
			const input = 'ȕʹ*椻򕰱麑ѩ+ֳ뢜1񒩎עqGǤ󻚩𑊵򼼊恳Ϛm슺Ĭꄩ󮭷񈥇󥷽ዯ-񆇳';
			const buffer = TextBufferView.fromString(input, 'utf8');
			expect(buffer.toString('utf8')).toBe(input);
			expect(buffer).toMatchSnapshot();
		});
	});

	it('creating from buffer', () => {
		const input = 'some random string';
		const buffer = Buffer.from(input, 'utf8');
		const textBuffer = TextBufferView.fromBuffer(buffer);
		expect(textBuffer.toString()).toBe(input);
		expect(textBuffer.originalArrayBuffer).toBe(buffer);
		expect(textBuffer.arrayBuffer).not.toBe(buffer);
		expect(textBuffer.arrayBuffer).not.toBe(textBuffer.originalArrayBuffer);
	});

	it('creating from array buffer', () => {
		const input = 'some random string';
		const buffer = new TextEncoder().encode(input);
		const textBuffer = TextBufferView.fromArrayBuffer(buffer);
		expect(textBuffer.toString()).toBe(input);
		expect(textBuffer.originalArrayBuffer).toBe(buffer);
		expect(textBuffer.arrayBuffer).not.toBe(buffer);
		expect(textBuffer.arrayBuffer).not.toBe(textBuffer.originalArrayBuffer);
	});

	describe('creating from non buffer fails', () => {
		it('data view', () => {
			const buffer = new ArrayBuffer(2);
			const view = new DataView(buffer);
			expect(() => {
				new TextBufferView(view as any);
			}).toThrowError('Invalid ArrayBuffer');
		})
	})
});
