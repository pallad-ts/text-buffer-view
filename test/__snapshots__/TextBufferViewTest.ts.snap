// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBufferView ascii should create from string 1`] = `
TextBufferView {
  "buffer": {
    "data": [
      97,
      115,
      99,
      105,
      105,
      32,
      99,
      104,
      97,
      114,
      115,
    ],
    "type": "Buffer",
  },
}
`;

exports[`TextBufferView base64 should create from string 1`] = `
TextBufferView {
  "buffer": {
    "data": [
      29,
      144,
      116,
      160,
      7,
      213,
      67,
      207,
      116,
      180,
    ],
    "type": "Buffer",
  },
}
`;

exports[`TextBufferView base64url should create from string 1`] = `
TextBufferView {
  "buffer": {
    "data": [
      159,
      103,
      252,
      108,
      17,
      217,
      16,
      95,
      159,
      235,
    ],
    "type": "Buffer",
  },
}
`;

exports[`TextBufferView hex should create from string 1`] = `
TextBufferView {
  "buffer": {
    "data": [
      96,
      205,
      172,
      139,
      57,
      56,
      241,
      136,
      237,
      96,
      28,
      202,
      70,
      21,
      1,
      150,
      218,
      31,
      241,
      196,
    ],
    "type": "Buffer",
  },
}
`;

exports[`TextBufferView utf8 should create from string 1`] = `
TextBufferView {
  "buffer": Uint8Array [
    200,
    149,
    205,
    180,
    42,
    230,
    164,
    187,
    242,
    149,
    176,
    177,
    233,
    186,
    145,
    209,
    169,
    238,
    139,
    145,
    43,
    214,
    179,
    235,
    162,
    156,
    49,
    241,
    146,
    169,
    142,
    215,
    162,
    113,
    71,
    199,
    164,
    243,
    187,
    154,
    169,
    240,
    145,
    138,
    181,
    242,
    188,
    188,
    138,
    230,
    129,
    179,
    207,
    154,
    109,
    236,
    138,
    186,
    196,
    172,
    234,
    132,
    169,
    243,
    174,
    173,
    183,
    241,
    136,
    165,
    135,
    243,
    165,
    183,
    189,
    225,
    139,
    175,
    45,
    241,
    134,
    135,
    179,
  ],
}
`;
